### PR TITLE
Add Log4J2LoggerFactory and Log4J2Logger

### PR DESCRIPTION
**Pull request by windie**
Motivation:

See #3095

Modifications:

Add Log4J2LoggerFactory and Log4J2Logger which is an InternalLogger implementation based on log4j2.

Result:

The user can use log4j2 directly without a special slf4j binding.

### CODE REVIEW

File: common/src/main/java/io/netty/util/internal/logging/Log4J2Logger.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+ */
+package io.netty.util.internal.logging;
+
+import org.apache.logging.log4j.Logger;
+
+public class Log4J2Logger extends AbstractInternalLogger {
+
+    private static final long serialVersionUID = 5485418394879791397L;
+
+    private final transient Logger logger;
```
Comment (netkins) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2Logger.java]: MINOR Make the "logger" logger private static final. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1312)

File: common/src/main/java/io/netty/util/internal/logging/Log4J2Logger.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util.internal.logging;
+
+import org.apache.logging.log4j.Logger;
+
+public class Log4J2Logger extends AbstractInternalLogger {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2Logger.java]: package private and final?

File: pom.xml [hunk 3]
```diff
@@ -383,6 +383,16 @@
         <version>1.1.3</version>
       </dependency>
       <dependency>
+        <groupId>org.apache.logging.log4j</groupId>
+        <artifactId>log4j-api</artifactId>
+        <version>2.3</version>
+      </dependency>
+      <dependency>
+        <groupId>org.apache.logging.log4j</groupId>
+        <artifactId>log4j-core</artifactId>
```
Comment (Scottmitch) [inline pom.xml]: does `core` belong here, or should this just be optional and/or for testing?

Comment (windie) [inline pom.xml]: It should be just for testing. Updated.

File: common/src/main/java/io/netty/util/internal/logging/Log4J2Logger.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+ */
+package io.netty.util.internal.logging;
+
+import org.apache.logging.log4j.Logger;
+
+final class Log4J2Logger extends AbstractInternalLogger {
+
+    private static final long serialVersionUID = 5485418394879791397L;
+
+    private final transient Logger logger;
```
Comment (netkins) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2Logger.java]: MINOR Make the "logger" logger private static final. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1312)

File: common/src/main/java/io/netty/util/internal/logging/Log4J2Logger.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+    protected Log4J2Logger(Logger logger) {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2Logger.java]: package private ?

File: common/src/main/java/io/netty/util/internal/logging/Log4J2LoggerFactory.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util.internal.logging;
+
+import org.apache.logging.log4j.LogManager;
+
+public class Log4J2LoggerFactory extends InternalLoggerFactory {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2LoggerFactory.java]: final ?

File: common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+import static org.easymock.EasyMock.verify;
+import static org.junit.Assert.assertTrue;
+
+public class Log4J2LoggerTest {
+    private static final Exception e = new Exception();
+
+    @Test
+    public void testIsTraceEnabled() {
+        Logger mock =
+            createStrictMock(Logger.class);
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java]: merge lines ?

File: common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+
+        InternalLogger logger = new Log4J2Logger(mock);
+        assertTrue(logger.isTraceEnabled());
+        verify(mock);
+    }
+
+    @Test
+    public void testIsDebugEnabled() {
+        Logger mock =
+            createStrictMock(Logger.class);
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java]: merge lines ?

File: common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+
+        InternalLogger logger = new Log4J2Logger(mock);
+        assertTrue(logger.isDebugEnabled());
+        verify(mock);
+    }
+
+    @Test
+    public void testIsInfoEnabled() {
+        Logger mock =
+            createStrictMock(Logger.class);
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java]: merge lines ?

File: common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+
+        InternalLogger logger = new Log4J2Logger(mock);
+        assertTrue(logger.isInfoEnabled());
+        verify(mock);
+    }
+
+    @Test
+    public void testIsWarnEnabled() {
+        Logger mock =
+            createStrictMock(Logger.class);
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java]: merge lines ?

File: common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+
+        InternalLogger logger = new Log4J2Logger(mock);
+        assertTrue(logger.isWarnEnabled());
+        verify(mock);
+    }
+
+    @Test
+    public void testIsErrorEnabled() {
+        Logger mock =
+            createStrictMock(Logger.class);
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java]: merge lines ?

File: common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+
+        InternalLogger logger = new Log4J2Logger(mock);
+        assertTrue(logger.isErrorEnabled());
+        verify(mock);
+    }
+
+    @Test
+    public void testTrace() {
+        Logger mock =
+            createStrictMock(Logger.class);
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java]: merge lines ?

File: common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+
+        InternalLogger logger = new Log4J2Logger(mock);
+        logger.trace("a", e);
+        verify(mock);
+    }
+
+    @Test
+    public void testDebug() {
+        Logger mock =
+            createStrictMock(Logger.class);
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java]: merge lines ?

File: common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+
+        InternalLogger logger = new Log4J2Logger(mock);
+        logger.trace("a");
+        verify(mock);
+    }
+
+    @Test
+    public void testTraceWithException() {
+        Logger mock =
+            createStrictMock(Logger.class);
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java]: merge lines ?

File: common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+
+        InternalLogger logger = new Log4J2Logger(mock);
+        logger.debug("a");
+        verify(mock);
+    }
+
+    @Test
+    public void testDebugWithException() {
+        Logger mock =
+            createStrictMock(Logger.class);
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java]: merge lines ?

File: common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+
+        InternalLogger logger = new Log4J2Logger(mock);
+        logger.debug("a", e);
+        verify(mock);
+    }
+
+    @Test
+    public void testInfo() {
+        Logger mock =
+            createStrictMock(Logger.class);
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java]: merge lines ?

File: common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+
+        InternalLogger logger = new Log4J2Logger(mock);
+        logger.info("a");
+        verify(mock);
+    }
+
+    @Test
+    public void testInfoWithException() {
+        Logger mock =
+            createStrictMock(Logger.class);
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java]: merge lines ?

File: common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+
+        InternalLogger logger = new Log4J2Logger(mock);
+        logger.info("a", e);
+        verify(mock);
+    }
+
+    @Test
+    public void testWarn() {
+        Logger mock =
+            createStrictMock(Logger.class);
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java]: merge lines ?

File: common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+
+        InternalLogger logger = new Log4J2Logger(mock);
+        logger.warn("a");
+        verify(mock);
+    }
+
+    @Test
+    public void testWarnWithException() {
+        Logger mock =
+            createStrictMock(Logger.class);
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java]: merge lines ?

File: common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+
+        InternalLogger logger = new Log4J2Logger(mock);
+        logger.warn("a", e);
+        verify(mock);
+    }
+
+    @Test
+    public void testError() {
+        Logger mock =
+            createStrictMock(Logger.class);
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java]: merge lines ?

File: common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+
+        InternalLogger logger = new Log4J2Logger(mock);
+        logger.error("a");
+        verify(mock);
+    }
+
+    @Test
+    public void testErrorWithException() {
+        Logger mock =
+            createStrictMock(Logger.class);
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/internal/logging/Log4J2LoggerTest.java]: merge lines ?

**Comment by normanmaurer** [comment]
@windie just a few comments...

@trustin any more comments ?

File: common/src/main/java/io/netty/util/internal/logging/Log4J2LoggerFactory.java [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util.internal.logging;
+
+import org.apache.logging.log4j.LogManager;
+
+public final class Log4J2LoggerFactory extends InternalLoggerFactory {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2LoggerFactory.java]: can this be a singleton?

Comment (windie) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2LoggerFactory.java]: I saw other similar classes are not singleton.

Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2LoggerFactory.java]: understood. I don't think we have to follow this "precedent" as the user could just use the singleton instead of creating a new object.

@normanmaurer - WDYT?

Comment (trustin) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2LoggerFactory.java]: They are internal classes, couldn't we just fix all of them?

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2LoggerFactory.java]: You mean make all final?

Comment (trustin) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2LoggerFactory.java]: Making them all final singleton.

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2LoggerFactory.java]: I think we should at least keep the public constructor to not break stuff as while its in the internal package user tend to set a logger

Comment (windie) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2LoggerFactory.java]: > Making them all final singleton.

I suggest that doing it in a separate PR since it's not related to the title.

In addition, should the constructor of these factory classes be private? If so, it will break some user codes since InternalLoggerFactory suggests using ` InternalLoggerFactory.setDefaultFactory(new Log4JLoggerFactory());`

Comment (trustin) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2LoggerFactory.java]: I wonder why a user calls `InternalLoggerFactory.setDefaultFactory()`. It tries to detect the right logger automatically in the order of following preference:

- SLF4J
- Log4J
- JUL

When does this fail and a user has to configure manually?

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2LoggerFactory.java]: I'm not sure if it ever fails but I have seen multiple times where user specify it manually (for whatever reason). So I would say let use mark the constructor deprecated and add static fields that should be used in the future. WDYT ?

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2LoggerFactory.java]: @trustin and I agree with @windie that we should just do this in another PR :+1:

Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2LoggerFactory.java]: new PR for fixing others sound good to me. we can discuss in the other PR but if these classes are internal lets just fix them ... i would rather not carry around deprecated APIs when we don't have to.

Comment (trustin) [inline common/src/main/java/io/netty/util/internal/logging/Log4J2LoggerFactory.java]: > @trustin[1] and I agree with @windie[2] that we should just do this in
> another PR :+1:
 
SGTM

Links:

  1. https://github.com/trustin
  2. https://github.com/windie

**Comment by windie** [comment]
Further comments?

**Comment by Scottmitch** [comment]
@windie - LGTM.

@normanmaurer - assigned to you ... do you want to pull in?

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (105df33d8d66701652ecc2159e7c8f01ad5d6a08) and 4.0 (35c5664a64125b0a2e32c56ae952c80a342b5ed7)
